crates/binstalk/src/helpers/git.rs

```diff
@@ -53,39 +89,87 @@ impl FromStr for GitUrl {
+        .with_shallow(remote::fetch::Shallow::DepthAtRemote(
+            NonZeroU32::new(1).unwrap(),
+        )))
+    }
+
     /// WARNING: This is a blocking operation, if you want to use it in
     /// async context then you must wrap the call in [`tokio::task::spawn_blocking`].
     ///
     /// WARNING: This function must be called after tokio runtime is initialized.
-    pub fn shallow_clone(url: GitUrl, path: &Path) -> Result<Self, GitError> {
-        let url_bstr = url.0.to_bstring();
-        let url_str = String::from_utf8_lossy(&url_bstr);
+    pub fn shallow_clone_bare(url: GitUrl, path: &Path) -> Result<Self, GitError> {
+        debug!("Shallow cloning {url} to {}", path.display());
 
-        debug!("Shallow cloning {url_str} to {}", path.display());
+        Ok(Self(
+            Self::prepare_fetch(url, path, create::Kind::Bare)?
+                .fetch_only(
+                    &mut TracingProgress::new(CompactString::new("Cloning")),
+                    &AtomicBool::new(false),
+                )?
+                .0
+                .into(),
+        ))
+    }
+
+    /// WARNING: This is a blocking operation, if you want to use it in
+    /// async context then you must wrap the call in [`tokio::task::spawn_blocking`].
+    ///
+    /// WARNING: This function must be called after tokio runtime is initialized.
+    pub fn shallow_clone(url: GitUrl, path: &Path) -> Result<Self, GitError> {
+        debug!("Shallow cloning {url} to {} with worktree", path.display());
 
         let mut progress = TracingProgress::new(CompactString::new("Cloning"));
 
         Ok(Self(
-            clone::PrepareFetch::new(
-                url.0,
-                path,
-                create::Kind::WithWorktree,
-                create::Options {
-                    destination_must_be_empty: true,
-                    ..Default::default()
-                },
-                open::Options::isolated(),
-            )?
-            .with_shallow(remote::fetch::Shallow::DepthAtRemote(
-                NonZeroU32::new(1).unwrap(),
-            ))
-            .fetch_then_checkout(&mut progress, &AtomicBool::new(false))?
-            .0
-            .main_worktree(&mut progress, &AtomicBool::new(false))?
-            .0,
+            Self::prepare_fetch(url, path, create::Kind::WithWorktree)?
+                .fetch_then_checkout(&mut progress, &AtomicBool::new(false))?
+                .0
+                .main_worktree(&mut progress, &AtomicBool::new(false))?
+                .0
+                .into(),
         ))
     }
+
+    #[inline(always)]
+    pub fn get_head_commit_entry_data_by_path(
+        &self,
+        path: impl AsRef<Path>,
+    ) -> Result<Option<Vec<u8>>, GitError> {
+        fn inner(this: &Repository, path: &Path) -> Result<Option<Vec<u8>>, GitError> {
+            Ok(
+                if let Some(entry) = this
+                    .0
+                    .to_thread_local()
+                    .head_commit()?
```

Comment:
The error handling is so much better now!

Reply:
Yeah, wrapping it in a function returns `GitError` removes all the `.map_err(GitError::from)`